dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -178,214 +177,422 @@ DDS::MemberId DynamicDataImpl::get_member_id_at_index(ACE_CDR::ULong index)
-    container_.single_map_.size() + container_.sequence_map_.size() + container_.complex_map_.size());
-  return count;
-}
-
-DDS::ReturnCode_t DynamicDataImpl::clear_all_values()
 {
   const TypeKind tk = type_->get_kind();
   switch (tk) {
-  case TK_ARRAY:
+  case TK_BOOLEAN:
+  case TK_BYTE:
+  case TK_UINT8:
+  case TK_UINT16:
+  case TK_UINT32:
+  case TK_UINT64:
+  case TK_INT8:
+  case TK_INT16:
+  case TK_INT32:
+  case TK_INT64:
+  case TK_FLOAT32:
+  case TK_FLOAT64:
+  case TK_FLOAT128:
+  case TK_CHAR8:
+#ifdef DDS_HAS_WCHAR
+  case TK_CHAR16:
+#endif
+  case TK_ENUM:
+    return 1;
+  case TK_STRING8:
+#ifdef DDS_HAS_WCHAR
+  case TK_STRING16:
+#endif
     {
-      DDS::TypeDescriptor_var td;
-      DDS::ReturnCode_t rc = type_->get_descriptor(td);
-      if (rc != DDS::RETCODE_OK) {
-        return rc;
+      if (!container_.single_map_.empty() || !container_.complex_map_.empty()) {
+        CORBA::ULong largest_index;
+        if (!container_.get_largest_index_basic(largest_index)) {
+          return 0;
+        }
+        return largest_index + 1;
+      }
+      return 0;
+    }
+  case TK_SEQUENCE: {
+    if (!container_.single_map_.empty() || !container_.complex_map_.empty()) {
+      CORBA::ULong largest_index;
+      if (!container_.get_largest_index_basic(largest_index)) {
+        return 0;
       }
-      const CORBA::ULong total = bound_total(td);
-      for (CORBA::ULong i = 0; i < total; ++i) {
-        rc = clear_value(get_member_id_at_index(i));
-        if (rc != DDS::RETCODE_OK) {
-          return rc;
+      if (!container_.sequence_map_.empty()) {
+        CORBA::ULong largest_seq_index;
+        if (!container_.get_largest_sequence_index(largest_seq_index)) {
+          return 0;
         }
+        largest_index = std::max(largest_index, largest_seq_index);
       }
-      return DDS::RETCODE_OK;
+      return largest_index + 1;
+    } else if (!container_.sequence_map_.empty()) {
+      CORBA::ULong largest_index;
+      if (!container_.get_largest_sequence_index(largest_index)) {
+        return 0;
+      }
+      return largest_index + 1;
     }
-    break;
+    return 0;
   }
-
-  if (tk == TK_UNION) {
-    const DDS::ReturnCode_t ret = reset_union();
-    if (ret != DDS::RETCODE_OK) {
-      return ret;
+  case TK_BITMASK:
+    return static_cast<ACE_CDR::ULong>(container_.single_map_.size() +
+                                       container_.complex_map_.size());
+  case TK_ARRAY: {
+    DDS::TypeDescriptor_var descriptor;
+    if (type_->get_descriptor(descriptor) != DDS::RETCODE_OK) {
+      return 0;
+    }
+    return bound_total(descriptor);
+  }
+  case TK_STRUCTURE: {
+    const CORBA::ULong member_count = type_->get_member_count();
+    CORBA::ULong count = member_count;
+    // An optional member that hasn't been set is considered missing.
+    // All non-optional members are counted since they either are set directly
+    // or hold default values.
```

Comment:
I'm quite sure I follow the part for testing if an optional member is set or not, but there is a check that if an optional member is not present in the container, then that means it hasn't been set by the user yet, and so is missing.

Reply:
I was meaning from the user perspective. For example if you got a DynamicData from a DynamicDataReader and an optional member was null. I'm wondering how to test for that because that could be significant. Maybe returning `RETCODE_NO_DATA` for those cases?